src/traces/sankey/render.js

```diff
@@ -216,27 +120,177 @@ function linkModel(uniqueKeys, d, l) {
+
+          // Walking back
+          'L' +
+          coords.sourceX + ' ' + (coords.sourceY + offset) +
+          'L' +
+          coords.leftInnerExtent + ' ' + (coords.sourceY + offset) +
+          'A' +
+          (coords.leftLargeArcRadius - offset) + ' ' + (coords.leftSmallArcRadius - offset) + ' 0 0 1 ' +
+          (coords.leftFullExtent - offset) + ' ' + (coords.sourceY + coords.leftSmallArcRadius) +
+          'L' +
+          (coords.leftFullExtent - offset) + ' ' + coords.verticalLeftInnerExtent +
+          'A' +
+          (coords.leftLargeArcRadius - offset) + ' ' + (coords.leftLargeArcRadius - offset) + ' 0 0 1 ' +
+          coords.leftInnerExtent + ' ' + (coords.verticalFullExtent - offset) +
+          'L' +
+          coords.rightInnerExtent + ' ' + (coords.verticalFullExtent - offset) +
+          'A' +
+          (coords.rightLargeArcRadius - offset) + ' ' + (coords.rightLargeArcRadius - offset) + ' 0 0 1 ' +
+          (coords.rightFullExtent + offset) + ' ' + coords.verticalRightInnerExtent +
+          'L' +
+          (coords.rightFullExtent + offset) + ' ' + (coords.targetY + coords.rightSmallArcRadius) +
+          'A' +
+          (coords.rightLargeArcRadius - offset) + ' ' + (coords.rightSmallArcRadius - offset) + ' 0 0 1 ' +
+          coords.rightInnerExtent + ' ' + (coords.targetY + offset) +
+          'L' +
+          coords.targetX + ' ' + (coords.targetY + offset) +
+          'Z';
+    }
+    return pathString;
+}
+
+function linkPath() {
+    var curvature = 0.5;
+    function path(d) {
+        if(d.link.circular) {
+            return createCircularClosedPathString(d.link);
+        } else {
+            var x0 = d.link.source.x1;
+            var x1 = d.link.target.x0;
+            var xi = interpolateNumber(x0, x1);
+            var x2 = xi(curvature);
+            var x3 = xi(1 - curvature);
+            var y0a = d.link.y0 - d.link.width / 2;
+            var y0b = d.link.y0 + d.link.width / 2;
+            var y1a = d.link.y1 - d.link.width / 2;
+            var y1b = d.link.y1 + d.link.width / 2;
+            return 'M' + x0 + ',' + y0a +
+                 'C' + x2 + ',' + y0a +
+                 ' ' + x3 + ',' + y1a +
+                 ' ' + x1 + ',' + y1a +
+                 'L' + x1 + ',' + y1b +
+                 'C' + x3 + ',' + y1b +
+                 ' ' + x2 + ',' + y0b +
+                 ' ' + x0 + ',' + y0b +
+                 'Z';
+        }
+    }
+    return path;
+}
+
+function nodeModel(d, n, i) {
     var tc = tinycolor(n.color);
     var zoneThicknessPad = c.nodePadAcross;
     var zoneLengthPad = d.nodePad / 2;
+    n.dx = n.x1 - n.x0;
+    n.dy = n.y1 - n.y0;
     var visibleThickness = n.dx;
     var visibleLength = Math.max(0.5, n.dy);
 
     var basicKey = n.label;
-    var foundKey = uniqueKeys[basicKey];
-    uniqueKeys[basicKey] = (foundKey || 0) + 1;
-    var key = basicKey + '__' + uniqueKeys[basicKey];
+    var key = basicKey + '__' + i;
```

Comment:
This key should always be unique regardless of the node's label because of `i`: it is the index of the node in the graph array and it will always be unique.

Please correct me if I'm wrong!

Reply:
That's sound ok to me :ok_hand: 

Resolved.